sky/serve/load_balancer.py

```diff
@@ -160,11 +160,26 @@ async def _proxy_with_retries(
         # SkyServe supports serving on Spot Instances. To avoid preemptions
         # during request handling, we add a retry here.
         retry_cnt = 0
+        # We keep track of the failed replicas for the current request, because
+        # we have a global round-robin policy, and if there is a large load,
+        # all retries for the same request can go to the same replica by chance.
+        # If the same replica is in `NOT_READY` state but the new state has not
+        # been synced from the controller, the current request will fail.
+        #
+        # We maintain a per-request failed replica set instead of the global
+        # one to allow multiple requests to still try failed replicas for one
+        # time, in case that replica is failed by transient network issue.
+        failed_replica_urls: Set[str] = set()
         while True:
             retry_cnt += 1
             with self._client_pool_lock:
+                # If all replicas are failed, clear the record and retry them
+                # again as some of them might be transient networking issues.
+                if (len(failed_replica_urls) ==
+                        self._load_balancing_policy.num_ready_replicas()):
```

Comment:
Just curious, is it possible that the `num_ready_replicas` updated when it is retrying in this fail loop, for example, the replicas scaled down. There should be concurrency issue here.

Should we instead check `if not (set(self._load_balancing_policy.ready_replicas) - failed_replica_urls): `